batch_builder.go

```diff
@@ -0,0 +1,228 @@
+	}
+
+	// replace_all
+	if _, err := b.buf.WriteString(`","replace_all":`); err != nil {
+		return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+	}
+	if _, err := b.buf.WriteString(boolString(b.replaceAll)); err != nil {
+		return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+	}
+
+	// ttl_minutes
+	if _, err := b.buf.WriteString(`,"ttl_minutes":`); err != nil {
+		return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+	}
+	if _, err := b.buf.WriteString(fmt.Sprintf("%d", b.ttlMinutes)); err != nil {
+		return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+	}
+
+	// service info, if set
+	if b.isSenderInfoSet() {
+		senderBytes, err := json.Marshal(b.sender)
+		if err != nil {
+			return nil, 0, fmt.Errorf("Error marshalling sender info to JSON: %s", err)
+		}
+
+		if _, err := b.buf.WriteString(`,"sender":`); err != nil {
+			return nil, 0, fmt.Errorf("Error writing sender info to buffer: %s", err)
+		}
+		if _, err := b.buf.Write(senderBytes); err != nil {
+			return nil, 0, fmt.Errorf("Error writing sender info to buffer: %s", err)
+		}
+	}
+
+	// upserts start
+	if _, err := b.buf.WriteString(`,"upserts":[`); err != nil {
+		return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+	}
+
+	// build a batch as big as we can
+	start := 0
+	end := len(b.serializedUpserts) - 1
+
+	// leave some space for batch scaffolding and `"complete":false`
+	availableSpace := b.desiredSize - b.buf.Len() - 40
+
+	upsertCount := 0
+	for {
+		if start > end || availableSpace <= 0 {
+			break
+		}
+
+		// try the bigger upserts first
+		if len(b.serializedUpserts[end]) <= availableSpace {
+			if upsertCount > 0 {
+				if _, err := b.buf.WriteString(","); err != nil {
+					return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+				}
+				availableSpace--
+			}
+			if _, err := b.buf.Write(b.serializedUpserts[end]); err != nil {
+				return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+			}
+			availableSpace -= len(b.serializedUpserts[end])
+
+			b.serializedUpserts[end] = nil
+
+			upsertCount++
+			end--
+			continue
+		}
+
+		// couldn't fit the bigger one - try smaller
+		if len(b.serializedUpserts[start]) <= availableSpace {
+			if upsertCount > 0 {
+				if _, err := b.buf.WriteString(","); err != nil {
+					return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+				}
+				availableSpace--
+			}
+			if _, err := b.buf.Write(b.serializedUpserts[start]); err != nil {
+				return nil, 0, fmt.Errorf("Error writing string to buffer: %s", err)
+			}
+			availableSpace -= len(b.serializedUpserts[start])
+
+			b.serializedUpserts[start] = nil
+
+			upsertCount++
+			start++
+			continue
+		}
+
+		break
+	}
+
+	if upsertCount == 0 {
+		return nil, 0, fmt.Errorf("Have %d remaining upserts, but could not fit any into the batch. The smallest one is %d bytes", len(b.serializedUpserts), len(b.serializedUpserts[start]))
```

Comment:
In tracing this code, I was thinking that the `hasClosedBatch` or `len(serializedUpserts) == 0` check would return quickly if there were no upserts added. However, it looks like if we have no upserts added then we will fail with a panic caused by this line, since there are no earlier checks and the `upsertCount` will be `0`. It may be good to tweak some things around so that if there are no serialized upserts it does one of the following:
- returns early with nil, 0, nil
- returns with a set of bytes in the json (but no upserts in this marshalled set of bytes)
- returns a specific error around the requirement to add an upsert
- something else?

Reply:
This is the unit test that I created but it panic'ed on:

```golang
// test behavior of batch builder when no upserts have been added
func TestBatchBuilder_Empty(t *testing.T) {
	a := require.New(t)

	maxSize := 10000
	sut := NewBatchBuilder(maxSize, true, 0)
	batchBytes, upsertCount, err := sut.BuildBatch()
	a.Nil(batchBytes)
	a.NoError(err)
	a.Zero(upsertCount)
}
```